### ISSUE TITLE
Generic instance exists for AnyRef

### ISSUE THREAD

**Issue by rhendric** [milessabin/shapeless#453]
`implicitly[shapeless.Generic[AnyRef]]` should fail, doesn't.

This also seems to affect other empty-constructor non-case classes (of which `java.lang.Object` is a special case). If I may editorialize on a fix, I'd propose adding a check to [`CaseClassMacros.isCaseClassLike.checkCtor`](https://github.com/milessabin/shapeless/blob/master/core/src/main/scala/shapeless/generic.scala#L575) requiring `params` to be non-empty, rather than just adding a guard for `AnyRef` specifically.

**Comment by milessabin** [comment]
I think your proposal is too drastic ... we definitely want to support case classes with empty constructors, case objects, `Unit` etc. The problem isn't so much the absence of contents, it's the availability of subtypes.

**Comment by rhendric** [comment]
I think actual case classes/objects/Unit would be fine; there's a check for actual case-class-ness on line 586, and a check for Unit in `isProduct`. This would only exclude non-case classes with empty constructors from being considered as empty products—and even those would have a second chance to be coproducts if they do have available subtypes, right?

**Comment by milessabin** [comment]
I still want to allow non-case classes with empty constructors to support [case classes a la carte](https://github.com/milessabin/shapeless/blob/master/examples/src/main/scala/shapeless/examples/alacarte.scala).

I'm rapidly coming to the conclusion that `AnyRef` really _is_ special. Other empty product types lack `AnyRef`'s near-universal supertype aspect, and I think that's just about enough to justify making a special exception.

**Comment by rhendric** [comment]
I agree that `AnyRef` is pretty special, and I don't think special-casing it is an awful idea. But I also think that having Generic support for empty-constructor classes can be a little dangerous.

A class with only an empty constructor is much more likely to have object identity semantics than value equality semantics. Take, as just an example, a custom Exception class that doesn't expose a message constructor. This CustomException and that CustomException are different things—they have different stack traces and so on. But if CustomException is used in a class that gets put through some scrap-your-boilerplate machinery, as far as Generic is concerned, that CustomException can be crushed into HNil and then rematerialized as a new CustomException with the wrong stack trace and so on. I'd much rather have a compiler error.

The power of scrap-your-boilerplate makes this an especially easy thing to overlook, too—it's hard to just follow the advice of the proverbial doctor who says ‘so don't do that’, because your identity-semantics-having class might be far removed from the root object that gets put into the machine, and it's very non-obvious that a failure of this kind is happening.

Granted, losing some out-of-band state like a stack trace is always a risk if you want to have a la carte case class functionality, and ordinarily I'd say that your trade-off in favor of a la carte is worth it (along with the good advice to use functional idioms instead of mutable state and object identity and so on). I'm arguing that in the specific case of empty constructor classes, the likelihood of out-of-band state goes way up (because there's no other way to supply state!), and the utility of those classes in an a la carte scheme goes way down (why wouldn't the a la carter just use an object instead of a class with an empty constructor?), and does that convince you to err on the side of safety for this case (and as a bonus it excludes the `AnyRef` problem)?

**Comment by milessabin** [comment]
Is it things like `CustomException` which are causing you particular trouble here? If it is, then I think we might be able to tighten up the rules to exclude them.

The intuition behind `Generic` is that it provides an isomorphism between a value of the target type and its representation type via a constructor (which creates a value from its parts) and a destructor (which extracts the parts from the value). It's not explicitly stated (it ought to be) but there is an expectation that the following will hold (at least contextually),

```scala
val gen = Generic[T]
val t1: T = ...
val t2: T = ...
val r1: gen.Repr = gen.to(t1)
val r2: gen.Repr = gen.to(t2)
assert(gen.from(r1) == t1)
assert(if(t1 == t2) r1 == r2 else true)
assert(if(r1 == r2) t1 == t2 else true)
```
Your `CustomException` doesn't support those assertions, not because it's empty, but because it gets information content which is relevant to equality from some route other than it's public primary constructor. Presumably that's via a private constructor? If so that suggests that one way of excluding it would be to bar every type which has any private constructors ... currently we only require that there is a matched pair of a constructor/destructor.

Would that (combined with a special case for `AnyRef`) cover the cases that are giving you trouble?

**Comment by rhendric** [comment]
It's not about private constructors; I was thinking of literally `class CustomException extends Exception("fixed message")`. The information content is the stack trace that `Exception` collects and stuffs in a field. There are other ways for classes to violate the above contract of `Generic`, of course; I wouldn't suggest you try to detect them all.

All I'm saying is that classes that only have a single, empty constructor (and just in case, I'm *not* talking about classes that have multiple constructors, one of which is empty; those are already refused by `Generic`) shouldn't be assumed to be isomorphic to the empty product, because they're `class`es and not singleton `object`s, which probably means there's some reason to have multiple unequal values of them. The mechanism by which the instances diverge isn't important; my comments about stack traces were just there to paint a picture of the impact of getting this wrong. And unless I'm grossly misunderstanding case classes a la carte (I might be! I've never tried them), users of that feature shouldn't be impacted by this choice (they can just use `object` for their singleton classes, unless they aren't really singletons but have no constructor parameters anyway, in which case aren't they doing fake case classes wrong?).

**Comment by milessabin** [comment]
Fixed in 287707228b24bcbde30e43b7b5a9d20c18fb0333.
